### PR TITLE
Support for --HEAD

### PR DESCRIPTION
**Pull request by ladislavb**
Implemented support for installation of unstable versions from git repository using --HEAD parameter

### CODE REVIEW

**Comment by ladislavb** [comment]
So far I have not found an easy way to install docbook-utils (jw...) on macOS so I copy
generated documentation from specified release - the survex version is hardcoded in the curl command. I was not able to figure out better way how to do that = it is probably not possible to use sed to parse version and pass it to other command as variable. Homebrew formula options seems to be quite limited in this area.

File: HomebrewFormula/survex.rb [hunk 1]
```diff
@@ -11,7 +12,21 @@ class Survex < Formula
...
+    depends_on "automake" => :build
+    depends_on "netpbm" => :build
+    depends_on "gnu-tar" => :build
+  end
+
   def install
+    if build.head?
+      system "autoreconf", "-fiv"
+      system "git", "checkout", "INSTALL"
+      system "curl https://survex.com/software/1.4.2/survex-1.4.2.tar.gz | gtar --strip-components=1 --skip-old-files -zxf -"
```
Comment (ojwb) [inline HomebrewFormula/survex.rb]: This will unpack everything from 1.4.2 over the top of the files from git, so will effectively just build 1.4.2.

Comment (ojwb) [reply HomebrewFormula/survex.rb]: Ah no, because of `--skip-old-files`....

**Comment by ojwb** [comment]
I think I'm going to leave this unmerged until we find a more satisfactory approach.

For the CI, taking various generated files from the last release is hack making a pragmatic trade-off.  It allows us to have useful CI for macos most of the time with some risk of it breaking sometimes when those files being out of step is actually a problem.  The alternative is not having any CI for macos (and so not spotting macos-specific breakage until end users are affected) until we can actually get a full build.  The downside of CI sometimes failing for bogus reasons is manageable - it's mostly me actually looking at CI results, and once I've determined a particular CI breakage on macos is bogus I can take that into account until it's resolved (which it should be by a new release), and potentially adjust that CI job temporarily to mark it as "known to fail" or even disable it.

But here this `--HEAD` option drops macos users into that can of worms, so the downside is more significant, and the upside less clear.  The real benefit I can see is it allows macos users to confirm a bug is actually fixed.

It does potentially provide a way for users to get a fix a little bit sooner (e.g. for the recent FFmpeg 5.0 compatibility fix), but I'll generally make a new release promptly once I'm aware of an issue that prevents Survex working on a particular platform.  In the FFmpeg case I fixed that in git on Feb 21 and the homebrew package was updated to include it on Feb 26, so only 5 days (and I wasn't aware that the FFmpeg 5.0 issue even affected homebrew at the point I fixed it - I fixed it because there was a debian bug open about it).

I suspect a `--HEAD` option might have the perverse effect of tending to increase the time to a released fix for macos issues such as this, since users may use `--HEAD` to work around them rather than actually reporting them.

But anyway, I don't think the hacky approach is really good enough for `--HEAD` - we don't want `--HEAD` to stop working because something incompatible changed with the files that are being taken from the previous release.

I'm planning to switch away from docbook - it wasn't a bad option when we started to use it, but it seems to be declining in popularly if anything, and at this point there are better documentation toolchains for our needs.  Probably sphinx-doc and restructured text, which should solve that aspect of this problem.

> Homebrew formula options seems to be quite limited in this area.

They're really just Ruby code - you should be able to do anything you can in Ruby.

**Comment by ladislavb** [comment]
Understood. I still think that --HEAD support is useful for:
1. testing of macOS related patches before public release
2. testing of major changes / translations etc. before public release

If I exclude all documentation I am able to build survex from git without taking various generated files from the release tarball. See forked version at https://github.com/ladislavb/homebrew-survex/blob/master/HomebrewFormula/survex.rb
It needs inkspace cask installed in advance to build iconset for Aven. Are you OK with that approach?

I tried to compile docbook-utils on macOS but it has a lot of other dependencies = not worth to spend time to create formula and maintain it. Especially if you plan to switch to something like reStructuredText or AsciiDoc or MarkDown with pandoc...

**Comment by ojwb** [comment]
So we can't list the inkscape cask as dependency?

I think that's a reasonable approach.  If we can't actually build-depend on inkscape it'd be good to give a clear error if it isn't found, but that can be handled by the `lib/icons/svg2iconsetzip` script (which would benefit developers on other platforms too).

**Comment by ojwb** [comment]
I've added that check to `lib/icons/svg2iconsetzip` and extended CI for this repo to test `--HEAD`.

I've also amended `brew install --HEAD` to `brew install survex --HEAD` in the README as I think the package name is required...

File: HomebrewFormula/survex.rb [hunk 2]
```diff
@@ -11,13 +12,32 @@ class Survex < Formula
...
+    end
+
     system "./configure", "--prefix=#{prefix}",
                           "--bindir=#{bin}",
                           "--mandir=#{man}",
                           "--docdir=#{doc}",
                           "--datadir=#{share}"
 
+    if build.head?
+      ENV.prepend_path "PATH", "/opt/homebrew/bin"
```
Comment (ojwb) [inline HomebrewFormula/survex.rb:37]: In the CI we fail to find inkscape.  The log shows:

```
 ==> Downloading https://media.inkscape.org/dl/resources/file/Inkscape-1.1.2.dmg
Warning: macOS's Gatekeeper has been disabled for this Cask
==> Installing Cask inkscape
==> Moving App 'Inkscape.app' to '/Applications/Inkscape.app'
==> Linking Binary 'inkscape.wrapper.sh' to '/usr/local/bin/inkscape'
🍺  inkscape was successfully installed!
```

So should we actually be adding `/usr/local/bin` to `PATH` here instead?

Comment (ojwb) [reply HomebrewFormula/survex.rb:37]: It seems this is a change in homebrew default for M1 macs vs x86 macs.  Trying to workaround...

Comment (ladislavb) [reply HomebrewFormula/survex.rb:37]: Yes, it is /opt/homebrew(bin on Apple Silicon (Arm) and /usr/local/bin on Intel. We can use `if Hardware::CPU.arm?` to run different commands on different CPU architectures. See https://rubydoc.brew.sh/Hardware/CPU.html

Comment (ojwb) [reply HomebrewFormula/survex.rb:37]: It's simpler to just add both directories to `PATH`.

The inkscape part is working now, but I can't seem to get `Locale::PO` to be found:

https://github.com/ojwb/homebrew-survex/runs/5546356970?check_suite_focus=true

Comment (ladislavb) [reply HomebrewFormula/survex.rb:37]: `Installing /usr/local/Cellar/perl/5.34.0/lib/perl5/site_perl/5.34.0/POD2/PT_BR/local/lib.pod`

It looks like you have installed Perl via homebrew. I am using Apple shipped version 5.30.3. It is quite outdated but it works. I had a lot of problems with homebrew Perl installation and it is really not recommended. If you need up to date version on macOS use https://perlbrew.pl

Comment (ojwb) [reply HomebrewFormula/survex.rb:37]: It's not what I've installed actually - it's installed by default in the github actions CI images for macos.

I've added `brew uninstall perl` so we're testing a more typical configuration, but that didn't fix it.

It seems homebrew runs formulae with `HOME` set to a temporary directory and no obvious way to know the real home directory, so I've changed things around to stop using `local::lib` and instead just install `Locale::PO` into that (temporary) home directory and point perl at it via `PERL5OPT`.

So `--HEAD` now works in CI, and should still for users (and the instructions are simpler as only `inkscape` needs to be installed by hand).

**Pull request merged**